File: frontend/src/lib/serviceAccounts.ts
import { apiGet, apiPost, apiPatch } from "@/lib/api";
import { isRecord } from "@/lib/typeGuards";

export type ServiceAccountToken = {
  id: string;
  service_account_id: string;
  token_prefix: string;
  status: string;
  scopes: string[];
  expires_at: string | null;
  last_used_at: string | null;
  created_at: string;
  revoked_at: string | null;
};

export type ServiceAccount = {
  id: string;
  name: string;
  description: string | null;
  enabled: boolean;
  scopes: string[];
  tokens: ServiceAccountToken[];
  created_at: string;
  updated_at: string;
};

export type CreateTokenResponse = {
  service_account: ServiceAccount;
  token: ServiceAccountToken;
  plaintext_token: string;
};

export function isServiceAccountToken(value: unknown): value is ServiceAccountToken {
  return (
    isRecord(value) &&
    typeof value.id === "string" &&
    typeof value.service_account_id === "string" &&
    typeof value.token_prefix === "string" &&
    typeof value.status === "string" &&
    Array.isArray(value.scopes) &&
    value.scopes.every((s) => typeof s === "string") &&
    (typeof value.expires_at === "string" || value.expires_at === null) &&
    (typeof value.last_used_at === "string" || value.last_used_at === null) &&
    typeof value.created_at === "string" &&
    (typeof value.revoked_at === "string" || value.revoked_at === null)
  );
}

export function isServiceAccount(value: unknown): value is ServiceAccount {
  return (
    isRecord(value) &&
    typeof value.id === "string" &&
    typeof value.name === "string" &&
    (typeof value.description === "string" || value.description === null) &&
    typeof value.enabled === "boolean" &&
    Array.isArray(value.scopes) &&
    value.scopes.every((s) => typeof s === "string") &&
    Array.isArray(value.tokens) &&
    value.tokens.every(isServiceAccountToken) &&
    typeof value.created_at === "string" &&
    typeof value.updated_at === "string"
  );
}

export async function listServiceAccounts(headers: HeadersInit): Promise<ServiceAccount[]> {
  const res = await apiGet<unknown>("/admin/service-accounts", { headers });
  if (!isRecord(res) || !Array.isArray(res.service_accounts)) {
    throw new Error("Invalid response format");
  }
  return res.service_accounts.filter(isServiceAccount);
}

export async function createServiceAccount(
  payload: { name: string; description: string | null; enabled: boolean },
  headers: HeadersInit,
): Promise<ServiceAccount> {
  const res = await apiPost<unknown>("/admin/service-accounts", payload, { headers });
  if (!isServiceAccount(res)) throw new Error("Invalid create response");
  return res;
}

export async function updateServiceAccount(
  id: string,
  payload: { name?: string; description?: string | null; enabled?: boolean },
  headers: HeadersInit,
): Promise<ServiceAccount> {
  const res = await apiPatch<unknown>(`/admin/service-accounts/${encodeURIComponent(id)}`, payload, { headers });
  if (!isServiceAccount(res)) throw new Error("Invalid update response");
  return res;
}

export async function createServiceAccountToken(
  id: string,
  payload: { scopes: string[]; expires_at: string | null },
  headers: HeadersInit,
): Promise<CreateTokenResponse> {
  const res = await apiPost<unknown>(`/admin/service-accounts/${encodeURIComponent(id)}/tokens`, payload, { headers });
  if (
    !isRecord(res) ||
    !isServiceAccount(res.service_account) ||
    !isServiceAccountToken(res.token) ||
    typeof res.plaintext_token !== "string"
  ) {
    throw new Error("Invalid create token response");
  }
  return {
    service_account: res.service_account as ServiceAccount,
    token: res.token as ServiceAccountToken,
    plaintext_token: res.plaintext_token,
  };
}

export async function revokeServiceAccountToken(
  tokenId: string,
  headers: HeadersInit,
): Promise<{ token_id: string; revoked: boolean }> {
  const res = await apiPost<unknown>(`/admin/service-account-tokens/${encodeURIComponent(tokenId)}/revoke`, {}, { headers });
  if (!isRecord(res) || typeof res.token_id !== "string" || typeof res.revoked !== "boolean") {
    throw new Error("Invalid revoke response");
  }
  return {
    token_id: res.token_id,
    revoked: res.revoked,
  };
}

File: backend/app/schemas/service_accounts.py
from __future__ import annotations

from typing import Any, Literal, TypeAlias

from pydantic import BaseModel, Field, field_validator

SERVICE_ACCOUNT_SCOPES = (
    "battle:create",
    "battle:read",
    "battle:execute",
    "vote:create",
)
_SERVICE_ACCOUNT_SCOPE_SET = frozenset(SERVICE_ACCOUNT_SCOPES)

ServiceAccountScope: TypeAlias = Literal[
    "battle:create",
    "battle:read",
    "battle:execute",
    "vote:create",
]
ServiceAccountTokenStatus: TypeAlias = Literal["active", "expired", "revoked"]


def _normalize_scopes(value: Any) -> Any:
    if value is None or not isinstance(value, list):
        return value

    normalized: set[str] = set()
    for scope in value:
        if not isinstance(scope, str):
            return value

        selected_scope = scope.strip()
        if not selected_scope:
            raise ValueError("Service-account scope cannot be empty")
        if selected_scope.startswith("admin:"):
            raise ValueError("Admin scopes are not supported for service accounts")
        if selected_scope not in _SERVICE_ACCOUNT_SCOPE_SET:
            raise ValueError(f"Unsupported service-account scope: {selected_scope}")

        normalized.add(selected_scope)

    return [scope for scope in SERVICE_ACCOUNT_SCOPES if scope in normalized]


class ServiceAccountTokenCreate(BaseModel):
    scopes: list[ServiceAccountScope] = Field(
        ...,
        min_length=1,
        max_length=len(SERVICE_ACCOUNT_SCOPES),
    )
    expires_at: str | None = None

    @field_validator("scopes", mode="before")
    @classmethod
    def _validate_scopes(cls, value: Any) -> Any:
        return _normalize_scopes(value)


class ServiceAccountTokenRedacted(BaseModel):
    id: str
    service_account_id: str
    token_prefix: str
    status: ServiceAccountTokenStatus = "active"
    scopes: list[ServiceAccountScope] = Field(default_factory=list)
    created_at: str
    expires_at: str | None = None
    last_used_at: str | None = None
    revoked_at: str | None = None

    @field_validator("scopes", mode="before")
    @classmethod
    def _validate_scopes(cls, value: Any) -> Any:
        return _normalize_scopes(value)


class ServiceAccountPublic(BaseModel):
    id: str
    name: str
    description: str | None = None
    enabled: bool
    scopes: list[ServiceAccountScope] = Field(default_factory=list)
    tokens: list[ServiceAccountTokenRedacted] = Field(default_factory=list)
    created_at: str
    updated_at: str

    @field_validator("scopes", mode="before")
    @classmethod
    def _validate_scopes(cls, value: Any) -> Any:
        return _normalize_scopes(value)


class ServiceAccountListResponse(BaseModel):
    service_accounts: list[ServiceAccountPublic]


class ServiceAccountCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=4096)
    enabled: bool = True
    scopes: list[ServiceAccountScope] = Field(
        default_factory=list,
        max_length=len(SERVICE_ACCOUNT_SCOPES),
    )

    @field_validator("scopes", mode="before")
    @classmethod
    def _validate_scopes(cls, value: Any) -> Any:
        return _normalize_scopes(value)


class ServiceAccountUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=4096)
    enabled: bool | None = None
    scopes: list[ServiceAccountScope] | None = Field(
        default=None,
        max_length=len(SERVICE_ACCOUNT_SCOPES),
    )

    @field_validator("scopes", mode="before")
    @classmethod
    def _validate_scopes(cls, value: Any) -> Any:
        return _normalize_scopes(value)


class ServiceAccountTokenCreateResponse(BaseModel):
    service_account: ServiceAccountPublic
    token: ServiceAccountTokenRedacted
    plaintext_token: str = Field(..., min_length=1)
